File: decision_trees/r/library.r
# Modeled off of "Extreme Gradient Boosting (XGBoost) with R and Python"
# Source: https://datascience-enthusiast.com/R/ML_python_R_part2.html

if(!require(readxl)) install.packages("readxl", repos = "http://cran.us.r-project.org")
if(!require(xgboost)) install.packages("xgboost", repos='http://cran.rstudio.com/')
if(!require(caret)) install.packages("caret", repos='http://cran.rstudio.com/')
library("readxl")
library("xgboost")
library("caret")

# Load and partition data.
dataset <- as.data.frame(read_excel("../../source/CCPP/Folds5x2_pp.xlsx"))
partition <- createDataPartition(y = dataset$PE, p = 0.75, list = FALSE)
train_data <- dataset[partition,]
test_data <- dataset[-partition,]

# Convert data into DMatrix format.
x_train = xgb.DMatrix(as.matrix(train_data[1:4]))
y_train = train_data$PE
x_test = xgb.DMatrix(as.matrix(test_data[1:4]))
y_test = test_data$PE

xgb_model = train(
  x_train, y_train,  
  method = "xgbTree"
)

# Calculate Mean Squared Error.
predicted = predict(xgb_model, x_test)
loss = (y_test - predicted)^2
error = sqrt(mean(loss))
cat('Mean Squared Error (MSE): ', error,'\n')

# Calculate R-squared.
y_test_mean = mean(y_test)
tss =  sum((y_test - y_test_mean)^2 )

residuals = y_test - predicted
rss =  sum(residuals^2)

r  =  1 - (rss/tss)
cat('R-squared: ', r, '\n')

# Write to file for plotting in Python.
losses <- c()
for (i in seq(1, by=1, length=length(y_test)))
	losses[i] = (y_test[i] - predicted[i])^2;

write.table(losses, file = "output_losses.txt", sep = "\n",
            row.names = FALSE)
write.table(y_test, file = "output_y_test.txt", sep = "\n",
            row.names = FALSE)

File: decision_trees/python/library.py
# Modeled off of "Gradient Boosting regression" on SciKit Learn (Library)
# Source: https://scikit-learn.org/stable/auto_examples/ensemble/plot_gradient_boosting_regression.html

import pandas as pd
import numpy as np
import matplotlib.pyplot as plt
import matplotlib.patches as mpatches
import utilities

from sklearn import ensemble
from sklearn.metrics import mean_squared_error

# Definition for Squared Sum of Errors.
def SSE(y_real, y_pred):
    sse = 0
    for index in range(len(y_real)):
        sse += (y_real[index] - y_pred[index]) ** 2

    return sse

# Definition for Squared Sum of Total Error.
def SST(y_real):
    sst = 0
    y_mean = np.mean(y_real)
    for index in range(len(y_real)):
        sst += (y_real[index] - y_mean) ** 2

    return sst

# Definition for R^2.
def R_squared(y_real, y_pred):
    return (1 - SSE(y_real, y_pred) / SST(y_real))

# Load dataset.
dataset = utilities.import_CCPP(False)
x = dataset.iloc[:,0:4]
y = dataset.iloc[:,4]
offset = int(x.shape[0] * 0.9)
x_train, y_train = x[:offset], y[:offset]
x_test, y_test = x[offset:], y[offset:]

# Fit regression model.
params = {'n_estimators': 500, 'max_depth': 4, 'min_samples_split': 2,
          'learning_rate': 0.01, 'loss': 'ls'}
clf = ensemble.GradientBoostingRegressor(**params)

clf.fit(x_train, y_train)
mse = mean_squared_error(y_test, clf.predict(x_test))
print("MSE: %.4f" % mse)

losses = (y_test.values.tolist() - clf.predict(x_test)) ** 2

# Plotting.
utilities.plot_test(losses, y_test.values.tolist(), 'Python-Library Loss for Test Dataset', 'palevioletred')
print(R_squared(y_test.values.tolist(), clf.predict(x_test).tolist()))
